app/api/csv/arrangeThesauri.ts

```diff
@@ -0,0 +1,180 @@
+    super(source.message);
+    this.source = source;
+    this.row = row;
+    this.index = index;
+  }
+}
+
+const separateSelectAndMultiselectThesauri = (
+  thesauriRelatedProperties: PropertySchema[] | undefined,
+  languages?: string[]
+): [{ [k: string]: string }, { [k: string]: string }] => {
+  const nameToThesauriIdSelects: { [k: string]: string } = {};
+  const nameToThesauriIdMultiselects: { [k: string]: string } = {};
+
+  thesauriRelatedProperties?.forEach(p => {
+    if (p.content && p.type) {
+      const thesarusID = p.content.toString();
+      if (p.type === propertyTypes.select) {
+        nameToThesauriIdSelects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdSelects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      } else if (p.type === propertyTypes.multiselect) {
+        nameToThesauriIdMultiselects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdMultiselects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      }
+    }
+  });
+
+  return [nameToThesauriIdSelects, nameToThesauriIdMultiselects];
+};
+
+type ThesauriValueData = {
+  thesauriIdToExistingValues: Map<string, Set<string>>;
+  thesauriIdToNewValues: Map<string, Set<string>>;
+  thesauriIdToNormalizedNewValues: Map<string, Set<string>>;
+};
+
+const setupIdValueMaps = (allRelatedThesauri: ThesaurusSchema[]): ThesauriValueData => {
+  const thesauriIdToExistingValues = new Map();
+  const thesauriIdToNewValues = new Map();
+  const thesauriIdToNormalizedNewValues = new Map();
+
+  allRelatedThesauri.forEach(t => {
+    if (t._id) {
+      const id = t._id.toString();
+      thesauriIdToExistingValues.set(
+        id,
+        new Set(t.values?.map(v => normalizeThesaurusLabel(v.label)))
+      );
+      thesauriIdToNewValues.set(id, new Set());
+      thesauriIdToNormalizedNewValues.set(id, new Set());
+    }
+  });
+
+  return { thesauriIdToExistingValues, thesauriIdToNewValues, thesauriIdToNormalizedNewValues };
+};
+
+const handleLabels = (
+  id: string,
+  original: string,
+  normalized: string | null,
+  thesauriValueData: ThesauriValueData
+) => {
+  if (
+    normalized &&
+    !thesauriValueData.thesauriIdToExistingValues.get(id)?.has(normalized) &&
+    !thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.has(normalized)
+  ) {
+    thesauriValueData.thesauriIdToNewValues.get(id)?.add(original);
+    thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.add(normalized);
+  }
+};
+
+const syncSaveThesauri = async (
+  allRelatedThesauri: ThesaurusSchema[],
+  thesauriIdToNewValues: Map<string, Set<string>>
+) => {
+  for (let i = 0; i < allRelatedThesauri.length; i += 1) {
+    const thesaurus = allRelatedThesauri[i];
+    if (thesaurus?._id) {
+      const newValues: { label: string }[] = Array.from(
+        thesauriIdToNewValues.get(thesaurus._id.toString()) || []
+      ).map(tval => ({ label: tval }));
+      if (newValues.length > 0) {
```

Comment:
since we are only saving thesaurus once, i do not think this gives us much performance gain, you can just save it and avoid a condition ?

Reply:
Moved condition to a filter prior to the loop.